### ISSUE TITLE
changing txpower

### ISSUE THREAD

**Issue by jmannt3ther** [RinCat/RTL88x2BU-Linux-Driver#27]
With this driver, no matter what we do the output of:

iw dev or iw wlan0 info

is: txpower 12.00 dBm

when trying to set: (after forcing down)
iw dev wlan0 set txpower fixed 20mBm or 30mBm, nothing changes.

The makefile has these options:

CONFIG_TXPWR_BY_RATE = y
CONFIG_TXPWR_BY_RATE_EN = y
CONFIG_TXPWR_LIMIT = y
CONFIG_TXPWR_LIMIT_EN = n

But they don't seem to change anything.

Has anyone been able to set the txpower to 20mBm or 30 or something other than 12?

**Comment by RinCat** [comment]
txpower is not implemented in the official driver.
https://github.com/RinCat/RTL88x2BU-Linux-Driver/blob/master/os_dep/linux/ioctl_cfg80211.c#L4180

**Comment by RinCat** [comment]
Lucky for you, I just updated to the last official driver I can found, and it "seems" support change TX power.
It can be set and read, but I have no way to test if it actually changed TX power.

Use `iw wlan0 set txpower fixed 2000` to change to 20.00 dBm.

**Comment by jmannt3ther** [comment]
Compiled and working. Seems it actually sets it on the adapter itself. I went back to the old driver on another OS and it was now 20dbm. Updating the driver definitely changed it from 12 to 20.  But this driver DOES let you change it. Thanks so much!

**Comment by jmannt3ther** [comment]
When running hostapd we are seeing:
wlan0: interface state UNINITIALIZED->COUNTRY_UPDATE
wlan0: interface state COUNTRY_UPDATE->HT_SCAN

**Comment by RinCat** [comment]
That seems normal, it need to check country regulations.

**Comment by jmannt3ther** [comment]
Hi, so after running a test with 2 clients overnight we got all these errors hostapd stopped working. Have you seen this before?

[dmesg_log.txt](https://github.com/RinCat/RTL88x2BU-Linux-Driver/files/5029126/dmesg_log.txt)

**Comment by RinCat** [comment]
No, but the code shows the error is H2C API command timeout. Is the device under heavy load?

**Comment by jmannt3ther** [comment]
At that time no, but sometimes yes. How can we prevent it? Are there any run-time flags or anything we can set?

**Comment by jmannt3ther** [comment]
I feel like this wasn't an issue in the older driver. Can you tell me where I can get your 5.6 version and do a comparison test?

**Comment by RinCat** [comment]
https://github.com/RinCat/RTL88x2BU-Linux-Driver/archive/314b662331971b10eb484fbcedb06bf5abfcb211.zip
Last 5.6 version.

**Comment by jmannt3ther** [comment]
Any ideas on those errors?

**Comment by jmannt3ther** [comment]
The manufacturer also provided me with an updated hostapd/wpa supplicant if that helps.

**Comment by RinCat** [comment]
No ideas, since I have never encountered this bug. You may try unload and load the driver again to see if it reset.
hostapd/wpa should be fine.

**Comment by jmannt3ther** [comment]
How about this error: RTW: ERROR rtl8822b_c2h_handler_no_io: C2H(9) enqueue FAIL!

**Comment by jmannt3ther** [comment]
How can we disable btrtl (bluetooth driver for same wireless card) and btusb for that card?

**Comment by RinCat** [comment]
> RTW: [HALMAC] [ALWAYS]shall R reg twice!!
> RTW: ERROR [HALMAC] [ERR] Dump efuse in suspend
is normal. But 
> RTW: ERROR rtl8822b_c2h_handler_no_io: C2H(9) enqueue FAIL!
is not.

**Comment by RinCat** [comment]
Bluetooth is disabled, or you have anything in `/proc/net/rtl88x2bu/wlan0/btcoex`?

**Comment by jmannt3ther** [comment]
its empty

**Comment by jmannt3ther** [comment]
we can recompile with CONFIG_BT_COEXIST = n instead of yes and try it also...

**Comment by RinCat** [comment]
You can give it a try

**Comment by jmannt3ther** [comment]
Any issue setting this in makefile?
CONFIG_POWER_SAVING = n

We also set:
iw dev wlan0 set power_save off

**Comment by RinCat** [comment]
you can turn it off

**Comment by jmannt3ther** [comment]
So today we have the following failure on the latest 5.8 driver:

`RTW: WARN _beamforming_leave: ` 

and then it deauth's the wifi device...and no longer allow anything to connect without a reboot.

**Comment by RinCat** [comment]
Set `rtw_beamform_cap=0` to disable beamform. But I think you probably crashed the device for some reason.
Make sure it has enough power supply and not overheat.

**Comment by jmannt3ther** [comment]
still getting these errors which break the ap...? any ideas?

[75756.600303] RTW: WARN _beamforming_leave: xx:xx:xx:xx:xx:xx is neither beamforming ee or er!!
[75756.600309] RTW: WARN _beamforming_leave: xx:xx:xx:xx:xx:xx is neither beamforming ee or er!!

**Comment by RinCat** [comment]
That means the mac address interface is not found.

**Comment by jmannt3ther** [comment]
mac for what? why does it break the ap?

**Comment by RinCat** [comment]
Maybe the interface disappeared for some reason. It cannot found an interface with the mac address.

**Comment by jmannt3ther** [comment]
Hi, so our adapter is usb 2.0 and no matter what modprobe command comes up it seems to be trying usb 3.0 mode. is there any other way to force usb 2.0 other than during modprobe? such as in the makefile?

**Comment by RinCat** [comment]
No, and the default is not switch unless you set `rtw_switch_usb_mode=1`.

**Comment by jmannt3ther** [comment]
In another bug #29 you reference a setting in the makefile that can enable usb 3.0. can you let us know what that is? our issues are related to when running under a virtual machine (virtualbox) the usb 3.0 xhci driver has errors. We need to disable usb 3.0. Thanks!

**Comment by RinCat** [comment]
There is no config in makefile, the place you can change USB version is use a [module parameter](https://github.com/RinCat/RTL88x2BU-Linux-Driver/blob/master/os_dep/linux/os_intfs.c#L391).
And the default is using USB 2.0 and does not upgrade to USB 3.0 unless you set the parameter.

**Comment by jmannt3ther** [comment]
OK awesome, we will check. Any other ideas related to xhci drivers? such as the option for CONFIG_USB_HCI?

**Comment by RinCat** [comment]
That is USB Host Controller Interface, nothing to do with USB3.0.

**Comment by jmannt3ther** [comment]
We are seeing this exact issue someone else is reporting when running the driver under virtualbox on alpine, under high load only. all other times it works great. 

https://www.virtualbox.org/ticket/19426

**Comment by RinCat** [comment]
It more likely a bug in the kernel, try update it.

**Comment by jmannt3ther** [comment]
Running the latest alpine offers. 5.x. Still have the same issues. Trying to come up with something.....

**Comment by RinCat** [comment]
Is your USB Controller ASMedia ASM1142?

**Comment by jmannt3ther** [comment]
No Intel Corporation 7 Series/C210 Series Chipset Family USB, i believe based on lshw

**Comment by jmannt3ther** [comment]
have you seen this issue?

RTW: ERROR rtl8822b_c2h_handler_no_io: C2H(9) enqueue FAIL!

**Comment by RinCat** [comment]
That is usually IO error

**Comment by jmannt3ther** [comment]
so we got our issues narrowed down. the only real issue we see is when there is heavy load we get these:

RTW:  ERROR _btmpoper_cmd: Didn't get h2C Rsp Event!

**Comment by RinCat** [comment]
Well, it means command timeout, may be the hardware overheated. You need to know this chip is kind of low-end, most of those adapters don't have a good heat dissipation efficiency.

**Comment by jmannt3ther** [comment]
I think that this might be caused by allowing hostapd to increase the tx power. It seems the default for the card is 12dbm. if we set a fixed value, it seems that hostapd is increasing it to 20dbm....how can we disable letting hostapd change it?

**Comment by jmannt3ther** [comment]
Also we are seeing this now:
wlan0: interface state HT_SCAN->DISABLED

**Comment by RinCat** [comment]
I don't think hostapd will change TX power, but you can set it after running hostapd.

interface state HT_SCAN->DISABLED is complex issues, can be cause by many things. you need more logs.

**Comment by jmannt3ther** [comment]
well i set it like this and its being changed by hosatpd:

iw wlan0 set txpower fixed 1200

**Comment by RinCat** [comment]
I don't think hosatpd allowed user set it manually.
Try `iw wlan0 set txpower limit 1200`?

**Comment by jmannt3ther** [comment]
is there a reason that after a while everything would just break?

**Comment by jmannt3ther** [comment]
also older drivers didn't set this, i think you added the ability for it to set dynamically, based on a previous request we had.

**Comment by jmannt3ther** [comment]
After setting txpower it shows as this:

![image](https://user-images.githubusercontent.com/38048656/90648614-bf477080-e207-11ea-815f-e2fb72dd00f0.png)

**Comment by jmannt3ther** [comment]
can we disable the power setting in makefile?

**Comment by RinCat** [comment]
No, the code don't change TX power unless something set it to the value. 
And you can't disable power setting in makefile, but you can commit it out 
https://github.com/RinCat/RTL88x2BU-Linux-Driver/blob/master/os_dep/linux/ioctl_cfg80211.c#L4357
Line 4357 to 4386.

**Comment by jmannt3ther** [comment]
So we are seeing everything work great but after a few hours it just stops, and nothing associates. I believe it to be a power save related issue.  Best way to disable in make file? Or also somewhere else?

we are seeing: 
![image](https://user-images.githubusercontent.com/38048656/91302557-af361080-e774-11ea-8787-7d812cd330d4.png)

**Comment by jmannt3ther** [comment]
Such as RTW_WOW_LPS_MODE=0

**Comment by RinCat** [comment]
There is a `CONFIG_POWER_SAVING` in Makefile

**Comment by jmannt3ther** [comment]
Have you seen this error before?
RTW: ERROR Free disconnecting network of scanned_queue failed due to pwlan == NULL

**Comment by RinCat** [comment]
No, but that means the target wifi network not exists.

**Comment by jmannt3ther** [comment]
Is there a good reference for these errors?

**Comment by RinCat** [comment]
The error is happened [here](https://github.com/RinCat/RTL88x2BU-Linux-Driver/blob/master/core/rtw_mlme.c#L1999). But don't have much information, only told you system can't find target wifi network.

**Comment by jmannt3ther** [comment]
Have you seen this? When we get this message, no clients can associate but hostapd is still running.

RTW: ERROR wait_halmac_event: Wait id(7, HALMAC_FEATURE_IQK) TIMEOUT! Reset HALMAC state!

**Comment by jmannt3ther** [comment]
Driver mounted, hostapd started, but then got that halmac error. Here is the full dmesg log. 

[dmesg_output.txt](https://github.com/RinCat/RTL88x2BU-Linux-Driver/files/5144434/dmesg_output.txt)

**Comment by jmannt3ther** [comment]
hostapd config

interface=wlan0
bridge=br0
driver=nl80211
ssid=1001
country_code=US
hw_mode=a
channel=161
#max_num_sta=255
auth_algs=3
wpa_group_rekey=0

ieee80211n=1
ieee80211ac=0
#wmm_enabled=1
#ieee80211d=1
#ieee80211h=1
#rts_threshold=-1
ignore_broadcast_ssid=0
ht_capab=[SHORT-GI-20][SHORT-GI-40][HT40+]

logger_syslog=-1
logger_syslog_level=2

wpa=2
wpa_passphrase=123456789
wpa_key_mgmt=WPA-PSK
#wpa_pairwise=TKIP
rsn_pairwise=CCMP
#macaddr_acl=0

**Comment by RinCat** [comment]
No not see it before, it still a kind of command timeout, not in the driver.

**Comment by jmannt3ther** [comment]
Does the driver support setting antenna diversity:

ant_sel=1 Or 2 when modprobing the driver? Or similar antenna options?

**Comment by RinCat** [comment]
`rtw_antdiv_cfg = 2` 

> 0:OFF , 1:ON, 2:decide by Efuse config

`rtw_antdiv_type = 0`

> 0:decide by efuse  1: for 88EE, 1Tx and 1RxCG are diversity.(2 Ant with SPDT), 2:  for 88EE, 1Tx and 2Rx are diversity.( 2 Ant, Tx and RxCG are both on aux port, RxCS is on main port ), 3: for 88EE, 1Tx and 1RxCG are fixed.(1Ant, Tx and RxCG are both on aux port)

`rtw_drv_ant_band_switch = 1`

> 0:OFF , 1:ON, Driver control antenna band switch

`rtw_single_ant_path = 0`

> 0:main ant , 1:aux ant , Fixed single antenna path, default main ant

By default, they all use firmware settings.

**Comment by jmannt3ther** [comment]
Any idea why no matter what the settings, the upload speed over wifi is always limited at 4-6Mbps? 
Download is 80Mbps+. The hostapd box itself can do 200/200Mbps+ up and down no issues.
We are seeing high ping times and low pings times and horrible upload speed. On different wifi channels as well.
Any ideas? Thanks!

**Comment by RinCat** [comment]
This may be the most common phenomenon of WiFi, channels bandwidths are shared by everyone. If you have some clients running at full speed, everyone will be slow. that not only limited to your clients.
And what is "upload", client to AP?

**Comment by jmannt3ther** [comment]
Yes client to AP. testing another kernel now.

**Comment by jmannt3ther** [comment]
The module we are using is actually an 8822bu not sure if that makes a difference

**Comment by RinCat** [comment]
8812bu and 8822bu use same driver.

**Comment by jmannt3ther** [comment]
The weird thing is that if i run the same card and antennas it works perfect with 80mbps+ up and down. Same hardware in linux and upload (client to ap) cant get past 7 or 8 mbps??

**Comment by jmannt3ther** [comment]
In windows 10 i meant to say

**Comment by RinCat** [comment]
maybe you can use [wavemon](https://github.com/uoaerg/wavemon) to see if your wifi something wrong.
I can reached 250+ Mbits/sec easily using iperf.

**Comment by jmannt3ther** [comment]
Wavemon doesn't show me any signal info. 

![image](https://user-images.githubusercontent.com/38048656/91665621-ef72f700-eac4-11ea-908c-ef28fed2904a.png)

**Comment by jmannt3ther** [comment]
look at the txpower:

`phy#0
        Interface wlan0
                ifindex 4
                wdev 0x1
                addr 00:15:61:52:b5:12
                type managed
                txpower 42949572.96 dBm
`

**Comment by jmannt3ther** [comment]
when hostapd is started the power level goes back to normal

**Comment by RinCat** [comment]
Try Wavemon at your client

**Comment by jmannt3ther** [comment]
Thanks, how about these errors:

[59572.722703] RTW: WARN _is_fw_read_cmd_down, reg_1cc(f), msg_box(0)...
[59572.722789] RTW: WARN  fw read cmd failed...

**Comment by RinCat** [comment]
Code comment said "pay attention to if race condition happened in H2C cmd setting".

**Comment by jmannt3ther** [comment]
does this driver support:

ignore_broadcast_ssid=1

does it work for you?

**Comment by RinCat** [comment]
`ignore_broadcast_ssid` is an internal parameter, used to hide the SSID and cannot be set externally.

**Comment by IMXSwarup** [comment]
> Lucky for you, I just updated to the last official driver I can found, and it "seems" support change TX power.
> It can be set and read, but I have no way to test if it actually changed TX power.
> 
> Use `iw wlan0 set txpower fixed 2000` to change to 20.00 dBm.

after reboot the tx power is returning to previous default one,, any solution?

**Comment by t3therdev** [comment]
Hi, old driver worked, same hardware. latest does this:

`
[    2.515683] input: PC Speaker as /devices/platform/pcspkr/input/input6
[    2.548538] Bluetooth: Core ver 2.22
[    2.548550] NET: Registered protocol family 31
[    2.548550] Bluetooth: HCI device and connection manager initialized
[    2.548559] Bluetooth: HCI socket layer initialized
[    2.548560] Bluetooth: L2CAP socket layer initialized
[    2.548563] Bluetooth: SCO socket layer initialized
[    2.554046] usbcore: registered new interface driver btusb
[    2.569092] Bluetooth: hci0: RTL: rtl: examining hci_ver=07 hci_rev=000b lmp_ver=07 lmp_subver=8822
[    2.569092]
[    2.571637] cfg80211: Loading compiled-in X.509 certificates for regulatory database
[    2.573360] Bluetooth: hci0: RTL: rom_version status=0 version=2
[    2.573360]
[    2.573361] Bluetooth: hci0: RTL: rtl: loading rtl_bt/rtl8822b_fw.bin
[    2.573361]
[    2.574143] bluetooth hci0: Direct firmware load for rtl_bt/rtl8822b_fw.bin failed with error -2
[    2.574145] Bluetooth: hci0: RTL: firmware file rtl_bt/rtl8822b_fw.bin not found
[    2.574145]
[    2.576712] cfg80211: Loaded X.509 cert 'sforshee: 00b28ddf47aef9cea7'
[    2.604403] 88x2bu: loading out-of-tree module taints kernel.
[    2.641122] RTW: module init start
[    2.641125] RTW: rtl88x2bu v5.13.1-20-gbd7c7eb9d.20210702_COEX20210316-18317b7b
[    2.641125] RTW: rtl88x2bu BT-Coex version = COEX20210316-18317b7b
[    2.641182] RTW: [HALMAC]55772M
[    2.641182] HALMAC_MAJOR_VER = 1
[    2.641182] HALMAC_PROTOTYPE_VER = 6
[    2.641182] HALMAC_MINOR_VER = 6
[    2.641182] HALMAC_PATCH_VER = 23
[    2.731454] input: ImExPS/2 Generic Explorer Mouse as /devices/platform/i8042/serio1/input/input7
[    5.684483] RTW: ERROR [HALMAC][ERR]Chip id is undefined
[    5.684485] RTW: ERROR rtw_halmac_init_adapter: halmac_init_adapter fail!(status=54)
[    5.684487] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.read_chip_version ###
[    5.684487] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.init_default_value ###
[    5.684488] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.intf_chip_configure ###
[    5.684488] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.read_adapter_info ###
[    5.684488] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_power_on ###
[    5.684489] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_power_off ###
[    5.684489] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_init ###
[    5.684489] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_deinit ###
[    5.684490] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.init_xmit_priv ###
[    5.684490] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.free_xmit_priv ###
[    5.684490] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_xmit ###
[    5.684491] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.mgnt_xmit ###
[    5.684491] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_xmitframe_enqueue ###
[    5.684491] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.init_recv_priv ###
[    5.684492] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.free_recv_priv ###
[    5.684492] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.inirp_init ###
[    5.684493] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.inirp_deinit ###
[    5.684493] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.dm_init ###
[    5.684493] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.dm_deinit ###
[    5.684494] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_dm_watchdog ###
[    5.684494] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.set_chnl_bw_handler ###
[    5.684494] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.set_hw_reg_handler ###
[    5.684495] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.GetHwRegHandler ###
[    5.684495] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.get_hal_def_var_handler ###
[    5.684495] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.SetHalDefVarHandler ###
[    5.684496] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.GetHalODMVarHandler ###
[    5.684496] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.SetHalODMVarHandler ###
[    5.684497] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.SetBeaconRelatedRegistersHandler ###
[    5.684497] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.fill_h2c_cmd ###
[    5.684497] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.hal_mac_c2h_handler ###
[    5.684498] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.fill_fake_txdesc ###
[    5.684498] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.fw_dl ###
[    5.684498] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.set_tx_power_level_handler ###
[    5.684499] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.set_tx_power_index_handler ###
[    5.684499] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.get_tx_power_index_handler ###
[    5.684499] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.sreset_init_value ###
[    5.684500] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.sreset_reset_value ###
[    5.684500] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.silentreset ###
[    5.684501] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.sreset_xmit_status_check ###
[    5.684501] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.sreset_linked_status_check ###
[    5.684501] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.sreset_get_wifi_status ###
[    5.684502] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.sreset_inprogress ###
[    5.684502] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.init_mac_register ###
[    5.684502] RTW: ### rtw_hal_ops_check - Error : Please hook hal_func.init_phy ###
[    5.684542] usbcore: registered new interface driver rtl88x2bu
`
